File: RecyclerAdapter.java
package com.oxylane.android.cubeinstore.ui.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.oxylane.android.cubeinstore.data.interfaces.IBaseCommunication;
import com.oxylane.android.cubeinstore.data.interfaces.IViewType;
import com.oxylane.android.cubeinstore.ui.fragment.interfaces.IAdapterChanged;
import com.oxylane.android.cubeinstore.ui.viewholder.RecyclerViewHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Maxime Jallu
 * @since 03/05/2017
 * <p>
 * Create for CubeInStore - Android (Decathlon)
 * <p>
 * Use this Class for : <br/>
 * ... {DOCUMENTATION}
 */

public class RecyclerAdapter<T> extends RecyclerView.Adapter<RecyclerViewHolder<T>> {

    private List<T> mTList;
    private ViewHolderFactory<T> mFactory;
    private IAdapterChanged mAdapterChanged;

    public RecyclerAdapter(ViewHolderFactory<T> factory){
        this(new ArrayList<>(), factory);
    }

    public RecyclerAdapter(List<T> TList, ViewHolderFactory<T> factory) {
        mTList = TList;
        mFactory = factory;
    }

    public void setFactory(ViewHolderFactory<T> factory) {
        mFactory = factory;
    }

    @Override
    public RecyclerViewHolder<T> onCreateViewHolder(ViewGroup parent, int viewType) {
        return mFactory.createVH(LayoutInflater.from(parent.getContext())
                .inflate(mFactory.getLayoutRes(viewType), parent, false), viewType);
    }


    @Override
    public void onBindViewHolder(RecyclerViewHolder<T> holder, int position) {
        holder.setItem(getItem(position));
        holder.setBound(false);
        holder.bind(holder.getItem());
        holder.setBound(true);
    }

    @Override
    public int getItemCount() {
        return mTList != null ? mTList.size() : 0;
    }

    /**
     * Get Item
     * @param position position founded
     * @return instance to position
     */
    public T getItem(int position){
        return mTList.get(position);
    }

    @Override
    public int getItemViewType(int position) {
        if(getItem(position) != null && getItem(position) instanceof IViewType){
            return ((IViewType) getItem(position)).getItemViewType();
        }
        return super.getItemViewType(position);
    }

    protected void callChangedListener(){
        if (mAdapterChanged != null){
            mAdapterChanged.onItemCountChange(getItemCount());
        }
    }

    public void setOnAdapterChangedListener(IAdapterChanged adapterChanged) {
        mAdapterChanged = adapterChanged;
    }

    protected void setCommunication(IBaseCommunication communication){
        mFactory.setCommunication(communication);
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     * @param item element to be inserted
     */
    public void addItem(T item){
        if (mTList != null) {
            mTList.add(item);
            notifyItemInserted(mTList.size());
        }
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     * @param item element to be inserted
     * @param position index at which the specified element is to be inserted
     */
    public void addItem(T item, int position){
        if (mTList != null) {
            position = Math.min(position, mTList.size());
            mTList.add(position, item);
            notifyItemInserted(position);
        }
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     * @param position the index of the element to be removed
     */
    public void removeItem(int position){
        if (mTList != null) {
            mTList.remove(position);
            notifyItemRemoved(position);
        }
    }

    /**
     * Set new list items and notifyDataSetChanged()
     * @link notifyDataSetChanged
     * @param list new instance items list for bind views
     */
    public void updateItems(@NonNull List<T> list){
        mTList = list;
        notifyDataSetChanged();
    }

    /**
     *
     * @return instance items list
     */
    public List<T> getTList() {
        return mTList;
    }
}

File: kotlin/RecyclerAdapter.kt
package {package_name}

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.oxylane.android.cubeinstore.data.interfaces.IBaseCommunication
import com.oxylane.android.cubeinstore.data.interfaces.IViewType
import com.oxylane.android.cubeinstore.ui.fragment.interfaces.IAdapterChanged
import com.oxylane.android.cubeinstore.ui.viewholder.RecyclerViewHolder
import java.security.AccessControlException
import java.util.*

/**
 * @author Maxime Jallu
 * @since 03/05/2017
 *
 * Use this Class for : <br></br>
 * ... {DOCUMENTATION}
 */
open class RecyclerAdapter<T> : RecyclerView.Adapter<RecyclerViewHolder<T>> {

    private var mTList: MutableList<T>
    private var mFactory: ViewHolderFactory<T>? = null
    private var mAdapterChanged: IAdapterChanged? = null

    constructor() {
        mTList = ArrayList()
    }

    constructor(factory: ViewHolderFactory<T>) : this(ArrayList<T>(), factory, null)

    constructor(viewHolderType: Class<out RecyclerViewHolder<T>>) : this(ArrayList<T>(), viewHolderType, null)

    constructor(viewHolderType: Class<out RecyclerViewHolder<T>>, callback: IBaseCommunication<*>?) : this(ArrayList<T>(), viewHolderType, callback)

    @JvmOverloads constructor(TList: MutableList<T>, viewHolderType: Class<out RecyclerViewHolder<T>>, callback: IBaseCommunication<*>? = null) : this(TList, ViewHolderFactory<T>(viewHolderType), callback)

    constructor(TList: MutableList<T>, factory: ViewHolderFactory<T>, callback: IBaseCommunication<*>?) {
        mTList = TList
        mFactory = factory
        mFactory!!.communication = callback
    }

    open fun setFactory(factory: ViewHolderFactory<T>) {
        mFactory = factory
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerViewHolder<T>? {
        if (mFactory == null) {
            throw AccessControlException("mFactory is not instancied. thanks use setFactory() method.")
        }

        return mFactory!!.createVH(LayoutInflater.from(parent.context)
                .inflate(mFactory!!.getLayoutRes(viewType), parent, false), viewType)
    }

    override fun onBindViewHolder(holder: RecyclerViewHolder<T>, position: Int) {
        holder.item = getItem(position)
        holder.isBound = false
        holder.bind(holder.item!!)
        holder.isBound = true
    }

    override fun getItemCount(): Int {
        return mTList.size
    }

    /**
     * Get Item
     *
     * @param position position founded
     * @return instance to position
     */
    fun getItem(position: Int): T {
        return mTList[position]
    }

    override fun getItemViewType(position: Int): Int {
        return if (getItem(position) != null && getItem(position) is IViewType) {
            (getItem(position) as IViewType).itemViewType
        } else super.getItemViewType(position)
    }

    fun putViewType(viewType: Int, viewHolder: Class<out RecyclerViewHolder<T>>) {
        mFactory!!.putViewType(viewType, viewHolder)
    }

    operator fun contains(obj: T): Boolean {
        return mTList.contains(obj)
    }

    protected fun callChangedListener() {
        if (mAdapterChanged != null) {
            mAdapterChanged!!.onItemCountChange(itemCount)
        }
    }

    fun setOnAdapterChangedListener(adapterChanged: IAdapterChanged) {
        mAdapterChanged = adapterChanged
    }

    fun setCommunication(communication: IBaseCommunication<*>) {
        mFactory!!.communication = communication
        notifyDataSetChanged()
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     *
     * @param item element to be inserted
     */
    fun addItem(item: T) {
        mTList.add(item)
        notifyItemInserted(mTList.size)
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     *
     * @param item     element to be inserted
     * @param position index at which the specified element is to be inserted
     */
    fun addItem(item: T, position: Int) {
        val pos = Math.min(position, mTList.size)
        mTList.add(pos, item)
        notifyItemInserted(pos)
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     *
     * @param collection elements to be inserted
     */
    fun addAll(collection: List<T>) {
        mTList.addAll(collection)
        val start = Math.max(0, mTList.size - collection.size - 1)
        notifyItemRangeInserted(start, collection.size)
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     *
     * @param item the element to be removed
     */
    fun removeItem(item: T) {
        removeItem(tList.indexOf(item))
    }

    /**
     * Inserts the specified element at the specified position in this list (optional operation).
     * Shifts the element currently at that position (if any) and any subsequent elements to the right (adds one to their indices).
     *
     * @param position the index of the element to be removed
     */
    fun removeItem(position: Int) {
        if (position > -1 && position < mTList.size) {
            mTList.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    /**
     * Set new list items and notifyDataSetChanged()
     *
     * @param list new instance items list for bind views
     * @link notifyDataSetChanged
     */
    fun updateItems(list: MutableList<T>) {
        mTList = list
        notifyDataSetChanged()
    }

    /**
     * @return instance items list
     */
    val tList: List<T>
        get() = mTList

    val isEmpty: Boolean
        get() = mTList.isEmpty()
}
